pkg/controller/machine/machine.go

```diff
@@ -463,36 +459,58 @@ func (c *Controller) ensureMachineHasNodeReadyCondition(machine *clusterv1alpha1
 	})
 }
 
-func (c *Controller) cleanupMachineAfterDeletion(machine *clusterv1alpha1.Machine) error {
-	var err error
-	glog.V(4).Infof("Removing finalizers from machine machine %s", machine.Name)
+// deleteMachine makes sure that an instance has gone in a series of steps.
+func (c *Controller) deleteMachine(prov cloud.Provider, machine *clusterv1alpha1.Machine) error {
+	if err := c.deleteCloudProviderInstance(prov, machine); err != nil {
+		c.recorder.Eventf(machine, corev1.EventTypeWarning, "DeletionFailed", "Failed to delete instance at cloud provider: %v", err)
+		return err
+	}
 
-	if machine, err = c.updateMachine(machine, func(m *clusterv1alpha1.Machine) {
-		finalizers := sets.NewString(m.Finalizers...)
-		finalizers.Delete(finalizerDeleteInstance)
-		finalizers.Delete(finalizerDeleteNode)
-		m.Finalizers = finalizers.List()
-	}); err != nil {
-		return fmt.Errorf("failed to update machine after removing the delete instance finalizer: %v", err)
+	// Delete the node object after the instance is gone
+	if sets.NewString(machine.Finalizers...).Has(finalizerDeleteInstance) {
+		return nil
+	}
+
+	if err := c.deleteNodeForMachine(machine); err != nil {
+		c.recorder.Eventf(machine, corev1.EventTypeWarning, "DeletionFailed", "Failed to delete node: %v", err)
+		return err
 	}
 
-	glog.V(4).Infof("Removed delete finalizer from machine %s", machine.Name)
 	return nil
 }
 
-// deleteMachine makes sure that an instance has gone in a series of steps.
-func (c *Controller) deleteMachine(prov cloud.Provider, machine *clusterv1alpha1.Machine) error {
-	if err := c.deleteProviderInstance(prov, machine); err != nil {
-		message := fmt.Sprintf("%v. Please manually delete %s finalizer from the machine object.", err, finalizerDeleteInstance)
-		c.recorder.Eventf(machine, corev1.EventTypeWarning, "DeletionFailed", "Failed to delete instance: %v", err)
-		return c.updateMachineErrorIfTerminalError(machine, common.DeleteMachineError, message, err, "failed to delete machine at cloudprovider")
+func (c *Controller) deleteCloudProviderInstance(prov cloud.Provider, machine *clusterv1alpha1.Machine) error {
+	finalizers := sets.NewString(machine.Finalizers...)
+	if !finalizers.Has(finalizerDeleteInstance) {
+		return nil
 	}
-	if err := c.deleteNodeForMachine(machine); err != nil {
-		message := fmt.Sprintf("%v. Please manually delete %s finalizer from the machine object.", err, finalizerDeleteNode)
-		c.recorder.Eventf(machine, corev1.EventTypeWarning, "DeletionFailed", "Failed to delete node: %v", err)
-		return errors.New(message)
+
+	// Retrieve the instance from the cloud provider
+	if _, err := prov.Get(machine); err != nil {
+		if err == cloudprovidererrors.ErrInstanceNotFound {
```

Comment:
I really dislike this change. This should be inside the provider code, it should simply block until it can't find the instance anymore.

Reply:
The deletion of instances is async for most providers we have(Not sure about vSphere) - thus its a more reliable implementation if we have that check above the providers.

Reply:
Okay, valid point. Can you please then add this info to the godoc of the interface definition? That the `Get` must return `cloudprovidererrors.ErrInstanceNotFound` when the instance does not exist?

Reply:
But actually this is nos that simple, because we must reprocess the machine  which will not happen by itself- Should we just rely on `resyncInterval` for that?

Reply:
The controller will reprocess the machine after 10 seconds max, see: https://github.com/kubermatic/machine-controller/pull/348/files#diff-87acff32b00826238809db10ac9ef562R391

Reply:
@mrIncompetent  Please add the expections regarding the `InstanceNotFound` err onto the godoc of the interface in `pkg/cloudprovider/cloud/provider.go`, then we are ready to go :)